src/object.jl

```diff
@@ -372,3 +372,147 @@ function Base.read(io::PrefetchedDownloadStream, ::Type{UInt8})
+        Base.@lock io.cond_wait begin
+            if length(io.eTags) < part_n
+                resize!(io.eTags, part_n)
+            end
+            io.eTags[part_n] = parteTag
+            io.ntasks -= 1
+            notify(io.cond_wait)
+        end
+    catch e
+        isopen(io.upload_queue) && close(io.upload_queue, e)
+        Base.@lock io.cond_wait begin
+            io.exc = e
+            notify(io.cond_wait, e, all=true, error=true)
+        end
+    end
+    return nothing
+end
+
+"""
+    MultipartUploadStream <: IO
+    MultipartUploadStream(args...; kwargs...) -> MultipartUploadStream
+
+An in-memory IO stream that uploads chunks to a URL in blob storage.
+
+Data chunks are written to a channel and spawned tasks read buffers from this channel.
+We expect the chunks to be written in order.
+Each task uploads a distinct part to blob storage.
+Each part is tagged with an id and we keep the tags into a vector. We also increment a
+counter to keep track of the number of parts. When there are no more data chunks to upload,
+we send a POST request with a single id for the entire upload.
+
+# Arguments
+* `store::AbstractStore`: The S3 Bucket / Azure Container object
+* `key::String`: S3 key / Azure blob resource name
+
+# Keywords
+* `credentials::Union{CloudCredentials, Nothing}=nothing`: Credentials object used in HTTP
+    requests
+* `concurrent_writes_to_channel::Int=(4 * Threads.nthreads())`: The number of concurrent
+    writes to the channel. Defaults to 4 times the number of threads. We use this value to
+    initialize a semaphore to perform throttling in case the writing to the channel is much
+    faster to uploading to blob storage.
+* `kwargs...`: HTTP keyword arguments are forwarded to underlying HTTP requests,
+
+## Examples
+```
+# Get an IO stream for a remote CSV file `test.csv` living in your S3 bucket
+io = MultipartUploadStream(bucket, "test.csv"; credentials)
+```
+
+## Performance
+We have benchmarked the performance of MultipartUploadStream for smaller (~39MB) and larger
+files up to ~860MB. For smaller files the performance is similar to an S3.put call, whereas
+for larger ones we do see a degradation of about 18% after ~300MB, that is growing more as
+the size of the uploaded file grows. We need to investirage further the cause of this.
+Some benchmark results can be found in https://github.com/JuliaServices/CloudStore.jl/pull/46#issuecomment-1804298709
+and https://github.com/JuliaServices/CloudStore.jl/pull/46#issuecomment-1810558208.
+"""
+mutable struct MultipartUploadStream{T <: AbstractStore} <: IO
+    store::T
+    url::String
+    credentials::Union{Nothing, AWS.Credentials, Azure.Credentials}
+    uploadState::Union{String, Nothing}
+    eTags::Vector{String}
+    upload_queue::Channel{Tuple{Int, Vector{UInt8}}}
+    cond_wait::Threads.Condition
+    cur_part_id::Int
+    ntasks::Int
+    exc::Union{Exception, Nothing}
+    sem::Base.Semaphore
+
+    function MultipartUploadStream(
```

Comment:
You can make sure people are using the right `credentials` for the right `store` by having two constructors:
```
MultipartUploadStream(store::AWS.Bucket, key::String; credentials::Union{Nothing, AWS.Credentials}, ...)
MultipartUploadStream(store::Azure.Container, key::String; credentials::Union{Nothing, Azure.Credentials}, ...)
```